cabal-install/Distribution/Client/ProjectPlanning.hs

```diff
@@ -688,12 +690,13 @@ getExecutableDBContents storeDirectory = do
     valid _ = True
 
 getSourcePackages :: Verbosity -> (forall a. (RepoContext -> IO a) -> IO a)
-                  -> Rebuild SourcePackageDb
-getSourcePackages verbosity withRepoCtx = do
+                  -> IndexUtils.IndexState -> Rebuild SourcePackageDb
+getSourcePackages verbosity withRepoCtx idxState = do
     (sourcePkgDb, repos) <-
       liftIO $
         withRepoCtx $ \repoctx -> do
-          sourcePkgDb <- IndexUtils.getSourcePackages verbosity repoctx
+          sourcePkgDb <- IndexUtils.getSourcePackagesWithState verbosity
+                                                               repoctx idxState
```

Comment:
So in future we could do better here.

Suppose a project is using `index-state: 123` and the user does `cabal update` somewhere else, so the source package db has changed, but the state of the source package db being used by this project has not (probably, unless the index was not just appended to but really changed -- e.g. as happened when we added the md5 checksums for all old packages).

So yes, ideally, normally we'd not force re-planning just because the index was appended to when it does not affect our project. We could do this by re-solving not on the change of the index but the change of the max timestamp or whatever it is from the index. That said, if we only used the timestamp then we would not notice if the content changed (as when the md5s were added), so it might be safer for index reading to return a checksum rather than just the final timestamp.

Just a thought for the future. It'd be a nice feature once we make index freezing the default workflow.